### ISSUE TITLE
URGENT: need script to remove spam [now fixed, but still needs some work]

### ISSUE THREAD

**Issue by remy** [jsbin/jsbin#2320]
So, here's the situation: spammers have filled the database with crud to 100% capacity. I woke up this morning at 6am to a few tweets and emails from users notifying me of this:

https://twitter.com/eggheadio/status/580925169671127041

I've got the start of a script that identifies spam users, but there's nothing consistent in the spammy bins (though usually they don't have any CSS or JS).

**The bigger problem is: I'm solo parenting my two young kids over the next 2 days and I don't have time to fix this. This sucks, but my kids come first.**

So, I'm calling on you - can you help? 

I reckon this is one approach:

1. Flag all spammy users (they come from a small subset of email domains) (script included below).
2. Get all bins from spammy users and delete them
3. Find bins that have HTML over a certain size... 2k? and no JS (so it's roughly fitting a profile) and has anonymous URL structure (`url.length > 6`) and remove.

This is a aggressive, but the disk space is at capacity and saving doesn't work at all right now.

Constraints:

1. The database is >15million records, so you can't just do a massive delete query, it needs to find records in chunks and needs to ideally remove individually.
2. User should be flagged with `UPDATE ownership SET flagged=1 WHERE name="bad-guy"`

Notes:

1. User table is called `ownership`
2. User bin tables is called `owners` - yes, I know, shit names
3. Bins are stored in `sandbox`

```js
var Sequelize = require('sequelize');
var sequelize = new Sequelize('database', 'root', 'password');

var spammy = ['spam-email.com'];

sequelize.query("SELECT * FROM `ownership`", { type: sequelize.QueryTypes.SELECT})
  .then(function(users) {
    var targets = users.filter(function (user) {
      var match = spammy.filter(function (spam) {
        return (user.email.indexOf('@' + spam) !== -1)
      });
      return match.length > 0;
    });

    targets.forEach(function (spam) {

    });

console.log(targets.length);
  })
```

**Comment by remy** [comment]
Few extra notes:

- Using mysql
- It probably makes sense to start gist or a repo to deal with this
- I've used `Sequelize` because it looked simple, but I don't care what's used

**Comment by remy** [comment]
If you have any questions, please ask here. I'll check in when I can to see if there's any progress or to run live scripts to get data back to the author.

**Comment by mathiasbynens** [comment]
jsPerf has the same problem.

* Cloudflare helps a bit (blocks traffic from known/detected spammers) but is not 100% foolproof.
* Implementing [Akismet](https://akismet.com/) didn’t really help.

**Comment by codecalm** [comment]
@remy you can update your database in one query:

```sql
UPDATE `ownership` SET flagged = 1 WHERE email REGEXP '\@(spam-email\.com|otherdomain\.com|other\.com)$'
```

this query set `flagged = 1` on each users in domains: spam-email.com, otherdomain.com, etc.

**Comment by michaelw90** [comment]
@remy Is there any ability to add capacity in the immediate interim? Or is it a fear of it filling up that is stopping you doing this?

I'll take a look when I get into work in about half an hour for you if it's not picked up

**Comment by sintaxi** [comment]
A few questions:

- So we're clear the system is currently FUBAR?
- when was your last backup?
- are you requiring email verification at any time?

We need a SQL master  in here. I'm going to see if I can get a hold of @silentrob.

**Comment by mathiasbynens** [comment]
@codecalm If the jsbin spam is anything like the jsPerf spam, it’s not that simple. The spammer uses various IP addresses, domain names, and URL shorteners.

**Comment by remy** [comment]
@codecalm this would work to flag the users. The `ownership` table is smaller, so it will run.

The next step is to identify the bins to remove them.

@mathiasbynens as a bandaid, I've removed "normal" registration and am only allowing github directed sign in. It should reduce the registered spam, but there's also a lot of anon spam that's hard to identify.

**Comment by codecalm** [comment]
@remy next step:

```sql
DELETE FROM sandbox WHERE user_id IN  (SELECT id FROM ownership WHERE flagged = 1)
```

**Comment by michaelw90** [comment]
Once the users are flagged, it's easy to then get their bins & delete them - you can just do a limit on the query - but run it a few times - this will limit your DB load.

**Comment by sintaxi** [comment]
@mathiasbynens thats useful to know. We can see this then as an arms race between the trouble it is to do those things and the impact that it has on the system.

**Comment by remy** [comment]
@michaelw90 the mounted volume is currently at 100%. Nothing is saving to disk anymore. I've started a snapshot process (it's hosted on AWS), and if I disconnect the EBS and resize it, it should alleviate the issue...a little), but I don't know what the timing is like.

**Comment by remy** [comment]
@codecalm that wouldn't work. The `sandbox` table is 15,000,000+ rows, lots of indices and uses indexdb. That would perform a full table scan which would kill the site entirely (probably for about 8 hours - this is based on the last table change I did a few years ago that locked the site for 4 hours).

That said, if that's the only option, then its the only option. One approach would be to do it in increment blocks (using `id BETWEEN a AND b`).

**Comment by remy** [comment]
@michaelw90 technically, yes, but the bulk of the spam doesn't come from the spam users (about 24,000 of them). It comes from anonymous bins, which is much, much hard to identify, but some rough rules might do the trick.

**Comment by remy** [comment]
@mathiasbynens a common use case for spammers on jsbin is the url shortener thing. They'll link to a bin on twitter or facebook, then that will bounce to their spam target via a short url and redirect in jsbin (most of which we automatically block these days).

**Comment by remy** [comment]
@codecalm ran your update to set the flag, and lol:

```text
ERROR 1114 (HY000): The table 'ownership' is full
```

I'm deleting a couple of known bins to make a few kb of space now.

**Comment by sintaxi** [comment]
It may be helpful to spin up a new instance of jsbin and switch over the DNS.

**Comment by codecalm** [comment]
@remy you can add `LIMIT xx` syntax into your `DELETE ...` query. Run it a few times and rows will be removed step by step.

**Comment by remy** [comment]
@sintaxi to what end? jsbin has two main servers: app + db. The EBS that stores the database on db is 100%, but app still works, which is why jsbin is still up (weirdly!)

**Comment by sintaxi** [comment]
Are you able to dump any data out?

**Comment by remy** [comment]
@sintaxi I can, do you mean to work with?

**Comment by codecalm** [comment]
@remy, so run this query few times:

```sql
UPDATE `ownership` SET flagged = 1 WHERE ownership = 0 AND email REGEXP '\@(spam-email\.com|otherdomain\.com|other\.com)$' LIMIT 1000
```

it update your database step by step.

**Comment by sintaxi** [comment]
@remy Im thinking rather than fixing the database. lets just get the good data out and stood up on a new machine.

**Comment by sintaxi** [comment]
@remy In other words fixing the problem isn't really necessary. Only the data is important. When was the last backup?

**Comment by remy** [comment]
@codecalm that worked for 1000 (ish) rows. Now `limit 1` doesn't work!

@sintaxi last backup was this morning. Aye, the data's important. Recovering could take a long long time, but yes, definitely my backup plan.

**Comment by sintaxi** [comment]
> last backup was this morning.

Thats great. Doesn't seem like a bad idea to me to spin up a new system and bootstrap with that backup. Then assess the loss.

**Comment by remy** [comment]
@sintaxi it's a manual job though (i.e. I don't have automation tools for jsbin's instances). And little'un wakes from nap very soon, so laptop goes down. Hence why I'm after bandaid to deal with junk bins.

**Comment by hugomallet** [comment]
Maybe remove sandbox before: (i have not verified the query)
```sql
DELETE sandbox FROM sandbox
LEFT JOIN ownership ON (sandbox.user_id = ownership.user_id)
WHERE ownership = 0 AND email REGEXP '\@(spam-email\.com|otherdomain\.com|other\.com)$';
```

**Comment by remy** [comment]
UPDATE: I've got a few minutes whilst my daughter naps, and my son is ipadding (not great). I'm going to unmount the volume, and create a new volume based on the last snapshot (this morning). Then remount.

I've no idea on the effect on jsbin if this happens. I suspect it'll be pretty bad, but nothing's saving right now anyway.

**Comment by remy** [comment]
@thisishugo that would take about 8+ hours to complete which is why I mentioned batching.

**Comment by michaelw90** [comment]
Do you have foreign keys & cascade deletes set up on your tables? 
If you do - would deleting a user then cascade through & delete their related bins?

**Comment by sintaxi** [comment]
> UPDATE: I've got a few minutes whilst my daughter naps, and my son is ipadding (not great). I'm going to unmount the volume, and create a new volume based on the last snapshot (this morning). Then remount.

Thats a good plan. :+1: 

> I've no idea on the effect on jsbin if this happens. I suspect it'll be pretty bad, but nothing's saving right now anyway.

Don't worry about that now.

**Comment by michaelw90** [comment]
@remy - when you restore, can you restore to a bigger EBS?

I can't remember with AWS - if you can dynamically resize an EBS - if so, can you just resize the one that is currently attached when you disconnect it?

**Comment by remy** [comment]
@michaelbragg no. I've not heard of cascade deletes before tbh.

Current status: new mount built based on snapshot - mounting and checking now.

**Comment by remy** [comment]
/mounting...

**Comment by matthiasbayer** [comment]
After you got a working instance back, it would be great to provide some additional data (access log, some database rows) so we can help you with finding a pattern and building a long term solution. 
Do you have any rate limiting yet? Are you using any DDoS protection service?

**Comment by remy** [comment]
ffs.

I've created a 400gb ebs (vol2), and mounted it on the data server, but (I assume) because it was created from a snapshot of an EBS that was 100GB (vol), it says that it's full:

```text
remy@db:/vol2$ df -kh
Filesystem      Size  Used Avail Use% Mounted on
/dev/xvdf        99G   94G  8.0K 100% /vol
/dev/xvdg        99G   94G  8.0K 100% /vol2
```

Ideas?

**Comment by remy** [comment]
@matthiasbayer I've got things like fail2ban in place which has helped a lot. I've also entirely disabled "free form" registration and only allow new users in via github (which should help a bit), but doesn't deal with general junk that lands.

**Comment by michaelw90** [comment]
Take a look at this @remy http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-expand-volume.html#recognize-expanded-volume-linux

**Comment by remy** [comment]
@michaelw90 the first part of that shows on my system that the EBS is the right size:

```text
remy@ip-10-168-9-52:~$ lsblk
NAME  MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
xvda1 202:1    0     8G  0 disk /
xvdb  202:16   0   410G  0 disk /mnt
xvdf  202:80   0   100G  0 disk /vol
xvdg  202:96   0   400G  0 disk
```

**Comment by remy** [comment]
Currently following these [directions](https://twitter.com/jdub/status/581023068245504000) and then [these](https://support.rightscale.com/12-Guides/Dashboard_Users_Guide/Clouds/AWS_Regions/EBS_Volumes/Actions/Grow_the_Size_of_an_EBS_Volume#Mount_and_Grow_the_Filesystem) if the first fail.

**Comment by michaelw90** [comment]
You should be able to resize `xvdf` using `resize2fs`

**Comment by remy** [comment]
Okay, mount correct size. Doing switherroo.

**Comment by remy** [comment]
Daughter expected to wake any minute now!!!

**Comment by remy** [comment]
Database is shutdown. Somehow jsbin.com is still running. Client side app ftw?.. ¯\_(ツ)_/¯

**Comment by remy** [comment]
Typical...I've forgotten how to start mysql on this machine. `mysqld` didn't do the trick. Shit. I wrote this down somewhere...

**Comment by michaelw90** [comment]
`service mysql start` most probably.

**Comment by remy** [comment]
Oh.

**Comment by dresende** [comment]
Yay!

**Comment by remy** [comment]
Back.

**Comment by matthiasbayer** [comment]
Look good from here

**Comment by remy** [comment]
Good job team. Same time next week?

**Comment by janl** [comment]
woot! 👏👏👏

**Comment by michaelw90** [comment]
Sure thing! :) :+1: 

I'd definitely look at getting your database updated to have foreign keys, with cascade deletes - so if a user deletes their account, their bins are automatically deleted etc.

**Comment by remy** [comment]
@michaelw90 to be honest, it's been a long outstanding item to completely ditch mysql and the database. The only thing that needs a database is the meta data (bin dates, etc), and the actual content could live on the file system. It's just a fucking massive job and you only have one shot.

**Comment by sintaxi** [comment]
Right On!! Nice work everyone.

**Comment by codecalm** [comment]
@remy optimize your database. If you have 15+ millions records set relations with foreign keys and cascade deletes. It's a one-time action and you avoid trouble in the future :)

**Comment by remy** [comment]
And all before this situation changed!
![2015-03-26](https://cloud.githubusercontent.com/assets/13700/6843687/2574a50a-d39d-11e4-94e9-b05a9ad8a165.jpg)

**Comment by codecalm** [comment]
:+1:

**Comment by mattseymour** [comment]
At least they are in a cage

**Comment by remy** [comment]
@mattseymour we like to call those cribs. Unlike the folks on MTV, who call houses, cribs. Which is just confusing...

**Comment by mattseymour** [comment]
:+1:

**Comment by remy** [comment]
Now the pressure is off, I'll find some time next week to purge flagged users' bins from the database (at least).

JS Bin *used* to automatically delete bins that were anonymous (in fact, all bins were anonymous) and idle for 3 months. The downside was that old sites that linked to real bins with low traffic would lose their content (but it wouldn't happen if they were signed in...). It might be something I restore (though perhaps a shorter time frame and the method it used was pretty poorly optimised).

**Comment by sage-oli-wood** [comment]
Just clicked though to this now so too late to help but it has been a joy to read.  Community++

**Comment by michaelw90** [comment]
@remy feel free to give me a shout if you want a hand with any DB optimisation or anything. :)

**Comment by janl** [comment]
@remy random idea from the armchair department: update the code so it writes new bins to disk and that viewing a bin can fetch from either DB or disk, so things at least don’t get a lot worse on the database end, and then some time later, migrate bins to disk one by one in the background and then clean up the database.

**Comment by SchizoDuckie** [comment]
Damn. wtf kind of satisfaction do people get out of creating spamscripts like these. I feel for you

**Comment by sage-oli-wood** [comment]
Shout if you'd like any views on how to monitor this sort of thing for the future, especially as it's AWS based.

**Comment by firasdib** [comment]
This is really unfortunate, and I'm sorry you're having to deal with all of this. But I don't think this is the correct forum to be discussing this, there's just too slow of a feedback loop. I suggest we talk through IRC. I am on freenode, and I'll be in `#savejsbin` from now on incase someone want to join and talk some ideas on how to resolve this.

**Comment by Mikola35** [comment]
:clap::clap::clap:

**Comment by silentrob** [comment]
Sorry I missed all the action last night. 

I also agree with @michaelw90, you want to get a foreign key constraint. This is super easy to do (and requires no code or app changes). When you identify the spammer, and delete the account, all attached bins are nuked - orphans are deleted.

The second thing you should do is add a rate limiter on creating bins (anonymous bins?) You could drop in https://github.com/tj/node-ratelimiter and that would probably fix the problem. 

If the bins are stored in SQL, you are probably right, they should be moved to DISK, but I wouldn't put them on an EBS disk, they have been known to disappear. S3 would be the right choice, and this would allow you to scale a little easier. Having said that, I still wouldn't replace MYSQL, you still need that join table back to the bin metadata anyway, and it is not worth completely re-writing. This might also help deal with mysql injection / html escaping issues - not that it is a problem, just a known attack vector.

You might also benefit from having a dead man switch and being able to though the site into read-only mode, but that would require a app level changes as well.

Lastly, if you want to attack the SPAM issue head on, there are tons of examples of creating a spam filter using Bayes Classification, you just need some labelled data for your training set. 
https://github.com/NaturalNode/natural#classifiers

Good luck and LMK if you need / want any help.

**Comment by perrybutler** [comment]
Couldn't you remove all records (bins) created on/after the date of incident as a simple rollback? I'd also suggest a rate limiter.

**Comment by NullEntity** [comment]
@remy What about deleting old bins after they haven't been accessed for 3 months (or some configurable amount of time)? You could store a timestamp on each bin. When a bin's loaded, update the timestamp. A cron task could run a simple statement to delete any bins with a last viewed timestamp greater than 3 months ago.

**Comment by NullEntity** [comment]
Ah, sorry, missed the comment where you mentioned that it used to do that.

**Comment by StoneCypher** [comment]
Can you give us an example of one of the spam pages?

**Comment by StoneCypher** [comment]
There may be some other steps we can take to rob the spammers of their motivation.  For other sites I would recommend decorating every link with `rel="nofollow"`, but given what you do that's probably not a practical answer.

With an example we might come up with something though

**Comment by mkdizajn** [comment]
community++ great to see that you jsbin is up! @remy ha :) my doughter has same (ikea) crib

**Comment by dcramer** [comment]
If you get my attention on any other day of the week I can tell you how to fix 99% of this. Writing things to disk is a fairly terrible idea fwiw. You can also improve the schema and usage as a 15m row table isn't too big and you should be able to run queries on it without such a headache. My GitHub name at gmail is probably the best way to get ahold of me if there's any interest.

**Comment by momomo** [comment]
Why not put a post timeout per session and/or ip? Maximum 1 save per 0.3 seconds or something?

**Comment by mathiasbynens** [comment]
@momomo https://github.com/jsbin/jsbin/issues/2320#issuecomment-86404045

**Comment by barneycarroll** [comment]
@rgrove had this problem with [rawgit](https://rawgit.com) and came up with [a clever algo for shutting down on abusive API spamming](https://medium.com/@yaypie/the-naughtiness-score-9a61872c61cd). Not perfect and not trivial to implement either, but it's a method…
